### DISCUSSION TITLE
Create if not exists

### DISCUSSION THREAD

**Discussion by kdev** [prisma/prisma#5815] [Q&A] [answered]
Is there a function that allows me to throw in create data and also checks if the `id` of the element to create already exists and if just returns that? I searched the docs but could only find `upsert`.

**Comment by nathanburgess** [comment, marked as answer]
`upsert` is what you want here. It'll create the record if it doesn't exist otherwise it will update it. Both behaviors will return the row.

**Reply by kdev** [reply]
Yeah but I still don't want to update a perfectly fine record?! Otherwise I'm just going to work around this but it's a feature I think is missing?!

**Reply by ruheni** [reply]
Hi @KokosnussDEV 👋🏽

I'm Alex from the Prisma team.

@nathanburgess suggested the right approach. Thank you for chiming in too.🙂

You can use `upsert`  to create a record. If the `update` property is empty, the record will not be updated.

Here's an example snippet:
```js
const upsertUser = await prisma.user.upsert({
  where: {
    email: 'viola@prisma.io',
  },
  update: {},
  create: {
    email: 'viola@prisma.io',
    name: 'Viola the Magnificent',
  },
})
```

We plan to document this better: how `upsert` can behave as a `findOrCreate`. The link to the GitHub issue can be found [here](https://github.com/prisma/docs/issues/640)

For more information regarding `upsert`, you can read more in the [Prisma Client API Reference](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#upsert) section.

**Reply by kdev** [reply]
Thanks! This was the information I was missing!

**Reply by piotros** [reply]
@ruheni is there any way to use foreing keys in `where` in `upsert`?

**Reply by ridhwaans** [reply]
what is the syntax to just perform a create in `upsert` if it doesnt exist and leave the where clause and update clause empty?

**Reply by ridhwaans** [reply]
also in the where clause if `id` is the only unique input and it is not known at the time of the query, it is a blocker

**Reply by ryands17** [reply]
> what is the syntax to just perform a create in upsert if it doesnt exist and leave the where clause and update clause empty?

If it's just empty you can assign `0` to the `id` as a workaround:

```ts
const upsertUser = await prisma.user.upsert({
  where: { id: id || 0 },
  update: {},
  create: {
    name: 'Name',
  },
})
```

> also in the where clause if id is the only unique input and it is not known at the time of the query, it is a blocker

`upsert` requires at least 1 unique value so you would need that to perform one otherwise it's no guarantee that a single record would be updated.

**Reply by ridhwaans** [reply]
I suppose using a composite id or `@@unique` will work

**Reply by mzw4** [reply]
This isn't working perfectly because it doesn't return the users that it finds. Is there a way for `upsert` to return the users even if it doesn't create or update them? Currently if it doesn't create or update, and just finds, it returns `{ count: 0 }`

**Reply by iiian** [reply]
I just came across this. Kind of frustrating that I cannot specify `where: { id: undefined }`. Semantically, that _feels_ pretty clear: "this mapping definitely doesn't exist, please create it." Instead, I get some very arcane error about `Available args are listed in green.`. I understand that instead, I can specify `where: { id: possibly_falsey_id_var ?? uuid.v4() }`, but why can't I delegate that responsibility :/

**Reply by wesamjabali** [reply]
To use different args in the where clause in prisma 3, define them as a unique set in your schema: 

```ts
model OrganizationUser {
  id             Int          @id @default(autoincrement())
  createdAt      DateTime     @default(now())
  user           User         @relation(fields: [userId], references: [id])
  organization   Organization @relation(fields: [organizationId], references: [id])
  userId         Int
  organizationId Int

  @@unique([userId, organizationId])
}
```

And then use it as your where clause like this:
```ts
this.prismaService.organizationUser.upsert({
    create: {
        user: { connect: { email } },
         organization: { connect: { id: organizationId } },
    },
    update: {},
    where: { userId_organizationId: { userId: user.id, organizationId } },
}),
```

**Reply by eniodev** [reply]
This will create a new record *if not exists* or update the current if it already exists. 
```js
return await prisma.user.upsert({
  where: { email: email },
  update: { ...user },
  create: { ...user },
});

**Reply by nathanburgess** [reply]
Please don’t respond to old threads with no new information. It sends an email to everyone subscribed.

**Reply by maxtaran2010** [reply]
You can do it like that sir

```js
return await prisma.user.upsert({
  where: { email: email },
  update: { ...user },
  create: { ...user },
});
```

**Reply by guigaribalde** [reply]
maybe doing something like this, sir:
```js
return await prisma.user.upsert({
  where: { email: email },
  update: { ...user },
  create: { ...user },
});
```

**Reply by tanvir0604** [reply]
What if I just want to create rows that doesn't exists without updating existing one? I can do it by checking in a loop, but is there any way to do it without loop?

**Reply by ghost** [reply]
for anyone interested in this. you can probably do it like this, sir:
```javascript
return await prisma.user.upsert({
  where: { email: email },
  update: { ...user },
  create: { ...user },
});
```

**Comment by michaelhays** [comment]
For anyone coming here for an example of how to achieve this with `upsert`, ruheni provides one in [this response](https://github.com/prisma/prisma/discussions/5815#discussioncomment-404334) (I wasn't able to see it by default because it was hidden under "Show previous replies")

**Comment by lasharela** [comment]
I have table of `Items`

```
model Items {
  id     String     @id @default(uuid())
  status ItemStatus @default(value: NEW)

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  tag      Tag    @relation(fields: [greenTag], references: [greenTag])
  greenTag String


  customer   Customer @relation(fields: [customerId], references: [id])
  customerId String
}
```

one tag can be added by multiple customers, so it's only unique for customers. 

When I'm trying to use upset, I'm not able to apply 

```
      where: {
        customerId: customer.id,
        greenTag,
      },
```
I can't apply customerID. But, when I'm trying to use `findFirst` all works perfectly:

```
    const tag = await db?.items.findFirst({
      where: {
        customerId: customer.id,
        greenTag,
      },
    });

```

What I'm doing wrong?

**Reply by janpio** [reply]
Please open a new discussion. This one is already marked as answered
